up-subscription/src/usubscription.rs

```diff
@@ -0,0 +1,560 @@
+/********************************************************************************
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use async_trait::async_trait;
+use log::*;
+use std::sync::Arc;
+use tokio::sync::{
+    mpsc::{self, UnboundedSender},
+    oneshot,
+};
+
+use crate::notification_manager::{self, NotificationEvent};
+use crate::{
+    helpers,
+    subscription_manager::{self, SubscriptionEvent},
+};
+
+use up_rust::core::usubscription::{
+    FetchSubscribersRequest, FetchSubscribersResponse, FetchSubscriptionsRequest,
+    FetchSubscriptionsResponse, NotificationsRequest, Request, SubscriptionRequest,
+    SubscriptionResponse, SubscriptionStatus, USubscription, UnsubscribeRequest,
+};
+use up_rust::{communication::RpcClient, LocalUriProvider, UCode, UStatus, UTransport, UUri};
+
+// Version used in USubscriptionService UEntity
+const USUBSCRIPTION_NAME: &str = "core.usubscription";
+
+/// Whether to include 'up:'  uProtocol schema prefix in URIs in log and error messages
+pub const INCLUDE_SCHEMA: bool = false;
+
+// Remote-subscribe operation ttl; 5 minutes in milliseconds, as per https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l3/usubscription/v3#6-timeout--retry-logic
+pub(crate) const UP_REMOTE_TTL: u32 = 300000;
+
+/// This trait (and the comprised UTransportHolder trait) is simply there to have a generic type that
+/// usubscription Listeners deal with, so that USubscriptionService can be properly mocked.
+pub trait USubscriptionServiceAbstract:
```

Comment:
Question on this. To implement any of the listeners, I had to pass in a "USubscriptionServiceAbstract" instead of a USubscriptionService. That was awkward, because I ended up having to add this line to usubscription.rs

// implement SubscriptionServiceAbstract
impl USubscriptionServiceAbstract for USubscriptionService {}

Is there a way for those listeners to take in the real service instead of abstract service?

Reply:
I am not 100% happy with the -Abstract thing myself, tbd - on the other hand, this is the best overall way I found so far to avoid having to deal with UTransport handles all over the place, while at the same time making all the mocks etc work properly. 

Let's agree to keep this one on the radar? I do expect some changes in this area anyways, when I'm picking up all the new base L2 implementations from up-rust...

Reply:
Yeah, sounds good to me :)